### PR TITLE
Adding kubemci e2e test for ingress spec conformance

### PR DESCRIPTION
**Pull request by nikhiljindal**
**What this PR does / why we need it**:
Adding an e2e test case for kubemci to verify that it conforms to the ingress spec.
Not all tests will pass right now, but adding it will enable us to track the latest status.

```release-note
NONE
```

### CODE REVIEW

**Comment by nikhiljindal** [comment]
This is still work in progress and needs some improvements but I wanted to send it early to get high level feedback on the direction.

cc @MrHohn @bowei @nicksardo @G-Harmon

**Comment by nikhiljindal** [comment]
Tracking kubemci access to this job in https://github.com/kubernetes/test-infra/issues/6624

File: test/e2e/framework/ingress_utils.go [hunk 1]
```diff
@@ -1051,10 +1051,34 @@ func (j *IngressTestJig) CreateIngress(manifestPath, ns string, ingAnnotations m
...
+	j.Ingress, err = j.RunCreate(j.Ingress)
 	ExpectNoError(err)
 }
 
+func (j *IngressTestJig) RunCreate(ing *extensions.Ingress) (*extensions.Ingress, error) {
+	if j.Class != MulticlusterIngressClassValue {
+		return j.Client.ExtensionsV1beta1().Ingresses(ing.Namespace).Create(ing)
+	}
+	// Use kubemci to create a multicluster ingress.
+	// TODO: Store the ingress in a file that can be passed to kubemci.
```
Comment (G-Harmon) [inline test/e2e/framework/ingress_utils.go]: please include username and/or Issue number for the TODOs

File: test/e2e/framework/ingress_utils.go [hunk 2]
```diff
@@ -1051,10 +1051,34 @@ func (j *IngressTestJig) CreateIngress(manifestPath, ns string, ingAnnotations m
...
+	filename := "ing-file"
+	// TODO: Add gcp-project
+	_, err := RunKubemciCmd("create", ing.Name, fmt.Sprintf("--ingress=%s", filename))
+	return ing, err
+}
+
+func (j *IngressTestJig) RunUpdate(ing *extensions.Ingress) (*extensions.Ingress, error) {
```
Comment (G-Harmon) [inline test/e2e/framework/ingress_utils.go]: Why do RunCreate and RunUpdate both call "kubemci create"? I think that makes sense, as a kubemci author, but it sure looks confusing as an outside reader.

File: test/e2e/framework/ingress_utils.go [hunk 3]
```diff
@@ -1051,10 +1051,34 @@ func (j *IngressTestJig) CreateIngress(manifestPath, ns string, ingAnnotations m
...
+	if j.Class != MulticlusterIngressClassValue {
+		return j.Client.ExtensionsV1beta1().Ingresses(ing.Namespace).Update(ing)
+	}
+	// Use kubemci to create a multicluster ingress.
```
Comment (G-Harmon) [inline test/e2e/framework/ingress_utils.go]: s/create/update/ ?

File: test/e2e/network/ingress.go [hunk 4]
```diff
@@ -520,6 +520,38 @@ var _ = SIGDescribe("Loadbalancing: L7", func() {
...
+		})
+
+		It("should conform to Ingress spec", func() {
+			jig.PollInterval = 5 * time.Second
+			conformanceTests = framework.CreateIngressComformanceTests(jig, ns, map[string]string{})
+			for _, t := range conformanceTests {
+				By(t.EntryLog)
+				t.Execute()
+				By(t.ExitLog)
+				jig.WaitForIngress(true)
```
Comment (G-Harmon) [inline test/e2e/network/ingress.go]: nit: document literals

**Review by G-Harmon — COMMENTED** [review]
looks great! sorry, I started to add nit-pick-y comments before remembering this is supposed to be high-level. I left them, but you can ignore for now or whatever.

**Review by MrHohn — COMMENTED** [review]
Thanks Nikhil! Adding the class field and extend the existing util functions seems right to me. In the future it may be good to have a `MultiIngressTestJig`, but this seems sufficient for now.

File: test/e2e/framework/ingress_utils.go (hunk 1 above)
Comment (nikhiljindal) [reply test/e2e/framework/ingress_utils.go]: Fixed it so its no longer valid

File: test/e2e/framework/ingress_utils.go (hunk 2 above)
Comment (nikhiljindal) [reply test/e2e/framework/ingress_utils.go]: Good point. Added a comment

File: test/e2e/framework/ingress_utils.go (hunk 3 above)
Comment (nikhiljindal) [reply test/e2e/framework/ingress_utils.go]: Done

File: test/e2e/network/ingress.go (hunk 4 above)
Comment (nikhiljindal) [reply test/e2e/network/ingress.go]: Done

**Comment by nikhiljindal** [comment]
Thanks for the review @G-Harmon and @MrHohn 
Updated code as per comments.
This is ready to be reviewed now.

Note that we need to fix https://github.com/kubernetes/test-infra/issues/6624 for the tests to run. I think it is fine to merge this, while we work on that. Let me know if you think otherwise.

File: test/e2e/framework/ingress_utils.go [hunk 5]
```diff
@@ -1057,10 +1058,39 @@ func (j *IngressTestJig) CreateIngress(manifestPath, ns string, ingAnnotations m
 		j.Ingress.Annotations[k] = v
 	}
 	Logf(fmt.Sprintf("creating" + j.Ingress.Name + " ingress"))
-	j.Ingress, err = j.Client.ExtensionsV1beta1().Ingresses(ns).Create(j.Ingress)
+	j.Ingress, err = j.RunCreate(j.Ingress)
 	ExpectNoError(err)
 }
 
+// RunCreate runs the required command to create the given ingress.
+func (j *IngressTestJig) RunCreate(ing *extensions.Ingress) (*extensions.Ingress, error) {
```
Comment (MrHohn) [inline test/e2e/framework/ingress_utils.go]: Could we simply name this `CreateIngress`?

File: test/e2e/framework/ingress_utils.go [hunk 6]
```diff
@@ -1057,10 +1058,39 @@ func (j *IngressTestJig) CreateIngress(manifestPath, ns string, ingAnnotations m
...
+	filePath := TestContext.OutputDir + "/mci.yaml"
+	if err := manifest.IngressToManifest(ing, filePath); err != nil {
+		return nil, err
+	}
+	_, err := RunKubemciCmd("create", ing.Name, fmt.Sprintf("--ingress=%s", filePath))
+	return ing, err
+}
+
+// RunUpdate runs the required command to update the given ingress.
+func (j *IngressTestJig) RunUpdate(ing *extensions.Ingress) (*extensions.Ingress, error) {
```
Comment (MrHohn) [inline test/e2e/framework/ingress_utils.go]: Name it `UpdateIngress`?

File: test/e2e/framework/ingress_utils.go [hunk 7]
```diff
@@ -1109,14 +1139,50 @@ func (j *IngressTestJig) GetRootCA(secretName string) (rootCA []byte) {
...
 // TryDeleteIngress attempts to delete the ingress resource and logs errors if they occur.
 func (j *IngressTestJig) TryDeleteIngress() {
-	err := j.Client.ExtensionsV1beta1().Ingresses(j.Ingress.Namespace).Delete(j.Ingress.Name, nil)
+	if err := j.RunDelete(j.Ingress, j.Class); err != nil {
+		Logf("Error while deleting the ingress %v/%v with class %s: %v", j.Ingress.Namespace, j.Ingress.Name, j.Class, err)
+	}
+}
+
+// RunDelete runs the required command to delete the given ingress.
+func (j *IngressTestJig) RunDelete(ing *extensions.Ingress, class string) error {
```
Comment (MrHohn) [inline test/e2e/framework/ingress_utils.go]: Name it `DeleteIngress`? BTW the second argument `class` is unused :)

File: test/e2e/manifest/manifest_test.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
+/*
```
Comment (MrHohn) [inline test/e2e/manifest/manifest_test.go]: Great to have unit test :)

**Review by MrHohn — APPROVED** [review]
Looks good, just couple minor comments

File: test/e2e/framework/ingress_utils.go (hunk 5 above)
Comment (nikhiljindal) [reply test/e2e/framework/ingress_utils.go]: There already exists a `CreateIngress` method . This is specific to running the Create command. CreateIngress calls RunCreate once it has computed the ingress to create.

File: test/e2e/framework/ingress_utils.go (hunk 6 above)
Comment (nikhiljindal) [reply test/e2e/framework/ingress_utils.go]: same as above

File: test/e2e/framework/ingress_utils.go (hunk 7 above)
Comment (nikhiljindal) [reply test/e2e/framework/ingress_utils.go]: same as above

File: test/e2e/manifest/manifest_test.go (hunk 8 above)
Comment (nikhiljindal) [reply test/e2e/manifest/manifest_test.go]: Yes! :)

File: test/e2e/framework/ingress_utils.go (hunk 5 above)
Comment (MrHohn) [reply test/e2e/framework/ingress_utils.go]: Oops, you are right, I scanned too quick. Maybe `RunCreateIngress`? Just wanted to indicate this is ingress specific :)

Comment (nikhiljindal) [reply test/e2e/framework/ingress_utils.go]: Its already a member function of IngressTestJig so RunCreate should be enough?
For creating other resources like services, we can name it explicitly RunCreateService.

I think RunCreate is good, but if you still feel strongly am happy to change it to RunCreateIngress.

**Comment by nikhiljindal** [comment]
Rebased

File: test/e2e/framework/ingress_utils.go [hunk 9]
```diff
@@ -1126,11 +1127,40 @@ func (j *IngressTestJig) CreateIngress(manifestPath, ns string, ingAnnotations m
 	for k, v := range ingAnnotations {
 		j.Ingress.Annotations[k] = v
 	}
-	j.Logger.Infof(fmt.Sprintf("creating" + j.Ingress.Name + " ingress"))
-	j.Ingress, err = j.Client.ExtensionsV1beta1().Ingresses(ns).Create(j.Ingress)
+	Logf(fmt.Sprintf("creating" + j.Ingress.Name + " ingress"))
```
Comment (MrHohn) [inline test/e2e/framework/ingress_utils.go]: Could we keep `j.Logger.Infof`?

File: test/e2e/framework/ingress_utils.go (hunk 5 above)
Comment (MrHohn) [reply test/e2e/framework/ingress_utils.go]: Yeah no worries, `RunCreate` should be enough

File: test/e2e/framework/ingress_utils.go [hunk 10]
```diff
@@ -1193,9 +1223,8 @@ func (j *IngressTestJig) TryDeleteIngress() {
 }
 
 func (j *IngressTestJig) TryDeleteGivenIngress(ing *extensions.Ingress) {
-	err := j.Client.ExtensionsV1beta1().Ingresses(ing.Namespace).Delete(ing.Name, nil)
-	if err != nil {
-		j.Logger.Infof("Error while deleting the ingress %v/%v: %v", ing.Namespace, ing.Name, err)
+	if err := j.RunDelete(ing, j.Class); err != nil {
+		Logf("Error while deleting the ingress %v/%v with class %s: %v", ing.Namespace, ing.Name, j.Class, err)
```
Comment (MrHohn) [inline test/e2e/framework/ingress_utils.go]: Use `j.Logger.Errorf` here

File: test/e2e/framework/ingress_utils.go [hunk 11]
```diff
@@ -1225,10 +1291,10 @@ func getIngressAddress(client clientset.Interface, ns, name string) ([]string, e
 }
 
 // WaitForIngressAddress waits for the Ingress to acquire an address.
-func (j *IngressTestJig) WaitForIngressAddress(c clientset.Interface, ns, ingName string, timeout time.Duration) (string, error) {
+func (j *IngressTestJig) WaitForIngressAddress(c clientset.Interface, ns, ingName, class string, timeout time.Duration) (string, error) {
```
Comment (MrHohn) [inline test/e2e/framework/ingress_utils.go]: We can use `j.Class` here now :)

**Review by MrHohn — APPROVED** [review]

File: test/e2e/framework/ingress_utils.go (hunk 11 above)
Comment (nikhiljindal) [reply test/e2e/framework/ingress_utils.go]: Done

**Comment by nikhiljindal** [comment]
Thanks for the quick replies!
Updated as per comments

File: test/e2e/framework/ingress_utils.go [hunk 12]
```diff
@@ -1127,10 +1128,39 @@ func (j *IngressTestJig) CreateIngress(manifestPath, ns string, ingAnnotations m
...
+func (j *IngressTestJig) RunCreate(ing *extensions.Ingress) (*extensions.Ingress, error) {
+	if j.Class != MulticlusterIngressClassValue {
+		return j.Client.ExtensionsV1beta1().Ingresses(ing.Namespace).Create(ing)
+	}
+	// Use kubemci to create a multicluster ingress.
+	filePath := TestContext.OutputDir + "/mci.yaml"
+	if err := manifest.IngressToManifest(ing, filePath); err != nil {
+		return nil, err
+	}
+	_, err := RunKubemciCmd("create", ing.Name, fmt.Sprintf("--ingress=%s", filePath))
```
Comment (madhusudancs) [inline test/e2e/framework/ingress_utils.go]: Is it necessary to export this function?

File: test/e2e/framework/ingress_utils.go [hunk 13]
```diff
@@ -1127,10 +1128,39 @@ func (j *IngressTestJig) CreateIngress(manifestPath, ns string, ingAnnotations m
...
+	return ing, err
+}
+
+// RunUpdate runs the required command to update the given ingress.
+func (j *IngressTestJig) RunUpdate(ing *extensions.Ingress) (*extensions.Ingress, error) {
```
Comment (madhusudancs) [inline test/e2e/framework/ingress_utils.go]: Same question here. Should these functions be internal?

File: test/e2e/manifest/manifest.go [hunk 14]
```diff
@@ -87,6 +91,20 @@ func IngressFromManifest(fileName string) (*extensions.Ingress, error) {
 	return &ing, nil
 }
 
+// IngressToManifest generates a yaml file in the given path with the given ingress.
+// Returns the file name and error, if there was any.
```
Comment (madhusudancs) [inline test/e2e/manifest/manifest.go]: It doesn't return the filename, does it?

File: test/e2e/manifest/manifest.go [hunk 15]
```diff
@@ -87,6 +91,20 @@ func IngressFromManifest(fileName string) (*extensions.Ingress, error) {
...
+func IngressToManifest(ing *extensions.Ingress, path string) error {
+	serialized, err := util.MarshalToYaml(ing, extensions.SchemeGroupVersion)
+	if err != nil {
+		return fmt.Errorf("failed to marshal ingress %v to YAML: %v", ing, err)
+	}
+
+	if err := ioutil.WriteFile(path, serialized, 0600); err != nil {
```
Comment (madhusudancs) [inline test/e2e/manifest/manifest.go]: What guarantees that that the parent directory exists? Is this expected to just error out in that case?

File: test/e2e/manifest/manifest_test.go [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+import (
+	"testing"
+
+	extensions "k8s.io/api/extensions/v1beta1"
+)
+
+func TestIngressToManifest(t *testing.T) {
+	ing := &extensions.Ingress{}
+	// Write the ingress to a file and ensure that there is no error.
+	if err := IngressToManifest(ing, "/tmp/ing.yaml"); err != nil {
```
Comment (madhusudancs) [inline test/e2e/manifest/manifest_test.go]: Do you want to use [`ioutil.TempDir()`](https://golang.org/pkg/io/ioutil/#TempDir)? Something like:

```go
tmpDir, err := ioutil.TempDir("", "mcie2e")
if err != nil {
    ...
}
defer os.RemoveAll(tmpDir)
if err := IngressToManifest(ing, filepath.Join(tmpDir, "ing.yaml")); err != nil {
   ...
}
```

Alternatively you can use [`ioutil.TempFile`](https://golang.org/pkg/io/ioutil/#TempFile) too.

File: test/e2e/manifest/manifest_test.go [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+		t.Fatalf("Error in creating file: %s", err)
+	}
+	// Writing it again should not return an error.
+	if err := IngressToManifest(ing, "/tmp/ing.yaml"); err != nil {
```
Comment (madhusudancs) [inline test/e2e/manifest/manifest_test.go]: Same here.

**Review by madhusudancs — COMMENTED** [review]
Minor nits. Feel free to either fix them in this PR or merge this PR and fix them in a follow up.

/lgtm
approve

File: test/e2e/manifest/manifest.go (hunk 14 above)
Comment (nikhiljindal) [reply test/e2e/manifest/manifest.go]: removed

File: test/e2e/manifest/manifest.go (hunk 15 above)
Comment (nikhiljindal) [reply test/e2e/manifest/manifest.go]: Yes, it assumes parent dir to exist. Added a comment for it

File: test/e2e/manifest/manifest_test.go (hunk 16 above)
Comment (nikhiljindal) [reply test/e2e/manifest/manifest_test.go]: Thanks.
This ensures that the test cleans up after it is run!

File: test/e2e/manifest/manifest_test.go (hunk 17 above)
Comment (nikhiljindal) [reply test/e2e/manifest/manifest_test.go]: done

File: test/e2e/framework/ingress_utils.go (hunk 13 above)
Comment (nikhiljindal) [reply test/e2e/framework/ingress_utils.go]: done

File: test/e2e/framework/ingress_utils.go (hunk 12 above)
Comment (nikhiljindal) [reply test/e2e/framework/ingress_utils.go]: made it private

**Review by nikhiljindal — COMMENTED** [review]
Thanks for the review @madhusudancs 
Sent https://github.com/kubernetes/kubernetes/pull/59964 and https://github.com/kubernetes/kubernetes/pull/59955 to fix the comments

**Pull request merged**